Test if PID is reused while NX window is open

rzinsly opened the pull request:
Adds a new test to check if a PID is being used
while a NX windows is still open with the same PID.

Signed-off-by: Raphael Moreira Zinsly <rzinsly@linux.ibm.com>

tuliom commented on test/test_pid_reuse.c:
```diff
@@ -0,0 +1,122 @@
+/* Test if children processes can reuse PID while NX window is open. */
```
Should this file have a licensing text?

tuliom commented on test/test_pid_reuse.c:
```diff
@@ -0,0 +1,122 @@
+        int err;
+        long *sleep_sec;
+
+	pid_t my_tid = syscall(SYS_gettid);
```
Wrong indentation?  This error appears on other lines too.

rzinsly replied:
Indeed, thanks for catching that.

tuliom commented on test/test_pid_reuse.c:
```diff
@@ -0,0 +1,122 @@
+		child_id = fork();
+
+		if (child_id > 0) {
+			if (child_id == parent_pid) {
```
Is this test targeting the kernel?
What is the relationship between fork() and libnxz?

rzinsly replied:
Yes, it is testing the kernel, what the test does is open an NX window and than fork other process to see if one of the child threads can use the same PID of the first one with the window still open.

I'll add more comments for the V2.

rzinsly replied:
We could use pthread_create instead of fork but I don't think it matters.

pmur replied:
I think I am missing something subtle.  When can the forked child's pid ever be the parent's pid?

rzinsly replied:
It shouldn't, there is a bug in kernel that a thread reuses a PID of another thread that is not running but left a NX window open, this test is checking if the system has a kernel with this problem.

lucmaga replied:
Is there already a fix for this kernel bug?

> It shouldn't, there is a bug in kernel that a thread reuses a PID of another thread that is not running but left a NX window open, this test is checking if the system has a kernel with this problem.

IMHO this explanation should be with the description on the file header. Also pointing 
out the kernel versions with this problem.

rzinsly replied:
> Is there already a fix for this kernel bug?

Yes

tuliom commented on test/test_pid_reuse.c:
```diff
@@ -0,0 +1,122 @@
+			if (child_id == parent_pid) {
+				fprintf(stderr, "!!! pid %d reused (see ps)\n",
+					child_id);
+				sleep(10000);
```
Why is this sleep necessary?
That's 10000 seconds.  Why not exiting?

rzinsly replied:
To keep this child alive until the point we get the csb error signal.

tuliom commented on test/test_pid_reuse.c:
```diff
@@ -0,0 +1,122 @@
+		} else if (child_id == 0) {
+			pid_t pid = getpid();
+			fprintf(stderr, "%d: child(%d)\n", i, (int) pid);
+			sleep(seconds);
```
Why is it sleeping 10 seconds?

rzinsly replied:
To give time to all children be up and not reuse the same PID of a closed child.

tuliom commented on test/test_pid_reuse.c:
```diff
@@ -0,0 +1,122 @@
+			exit(0);
+		} else if (child_id == -1) {
+			fprintf(stderr, "fork failed: %s\n", strerror(errno));
+			sleep(2);
```
Likewise...

rzinsly replied:
In fact here we could exit.

rzinsly commented:
I rebased the patch with the current develop branch, fixed indentation, added more comments and updated the sleep for when the PID is reuse. The time intended was never 10000 but 1000 to be around 15 minutes, but in the system I tested 5 minutes is enough time to catch it.

pmur commented on test/test_pid_reuse.c:
```diff
@@ -0,0 +1,166 @@
+	int children = 2048;
+
+	for (i=0; i < children; i++) {
+		child_id = fork();
```
Is this explicitly using a double-fork trick to avoid zombies?

rzinsly replied:
No, it is forking 2048 times to see if any child uses the same pid as the parent.

pmur commented on test/test_pid_reuse.c:
```diff
@@ -0,0 +1,166 @@
+	long *sleep_sec;
+
+	pid_t my_tid = syscall(SYS_gettid);
+	fprintf(stderr, "thread started: pid: %d\n", (int) my_tid);
```
Should the message read tid?

rzinsly commented on test/test_pid_reuse.c:
```diff
@@ -0,0 +1,166 @@
+		iterations = 100;
+
+	/* Get parent thread PID. */
+	pid_t my_tid = syscall(SYS_gettid);
```
We are looking at the kernel PID, getpid returns the user PID that is the same for all threads in the same process.

tuliom commented on test/test_pid_reuse.c:
```diff
@@ -0,0 +1,167 @@
+	/* Fork many children to see if any use the same PID as the parent. */
+	fork_children((int) my_pid);
+	/* Wait for all children to run. */
+	sleep(60);
```
If the goal is to wait for all children processes, you can use `wait()`, calling it in a loop until it returns -1 and `errno == ECHILD`.

rzinsly replied:
Does it triggers when the child is sleeping?
Because we need to continue while the child that reused the PID is sleeping, not killed. (line 85)

tuliom replied:
> Does it triggers when the child is sleeping?

@rzinsly No, it does not.

> we need to continue while the child that reused the PID is sleeping, not killed. (line 85)

In that case, I'm afraid there is a bug in this code.
The parent could exit before all its children complete without calling `wait()`, leaving zombie processes behind.

rzinsly replied:
It shouldn't 60 seconds proved enough to wait for all children to finish or to get to the point that we hit the bug, in this case it will segfault later.
I added the wait() in loop at the end of the test for safety anyway.

pmur commented on test/test_pid_reuse.c:
```diff
@@ -0,0 +1,172 @@
+	else
+		iterations = 100;
+
+	/* Get parent thread kernel PID. */
```
Should this read "Get parent thread ID"?  Similarly renaming the variable.

rzinsly replied:
@pmur, we are looking for a kernel problem, I believe it's better to use the kernel naming and call it PID instead of TID.
